importer,workload: reduce allocs

dt opened the pull request:
See commits.
```
benchstat before.txt after.txt
goos: darwin
goarch: arm64
                       │ before.txt │             after.txt             │
                       │   sec/op   │   sec/op     vs base              │
ConvertToKVs_TPCC01-10   2.025 ± 1%   1.988 ± 17%       ~ (p=0.065 n=6)
ConvertToKVs_TPCC10-10   18.72 ± 3%   18.22 ±  3%  -2.63% (p=0.002 n=6)
geomean                  6.156        6.019        -2.23%

                       │  before.txt  │              after.txt              │
                       │     B/s      │      B/s       vs base              │
ConvertToKVs_TPCC01-10   66.53Mi ± 1%   67.77Mi ± 14%       ~ (p=0.065 n=6)
ConvertToKVs_TPCC10-10   66.21Mi ± 3%   68.00Mi ±  3%  +2.70% (p=0.002 n=6)
geomean                  66.37Mi        67.88Mi        +2.28%

                       │   before.txt   │              after.txt               │
                       │      B/op      │     B/op       vs base               │
ConvertToKVs_TPCC01-10   1277.9Mi ± 18%   950.9Mi ± 24%  -25.59% (p=0.002 n=6)
ConvertToKVs_TPCC10-10   12.366Gi ±  4%   8.967Gi ±  4%  -27.49% (p=0.002 n=6)
geomean                   3.928Gi         2.886Gi        -26.54%

                       │ before.txt  │             after.txt              │
                       │  allocs/op  │  allocs/op   vs base               │
ConvertToKVs_TPCC01-10   22.65M ± 7%   18.22M ± 8%  -19.53% (p=0.002 n=6)
ConvertToKVs_TPCC10-10   212.5M ± 2%   172.8M ± 2%  -18.69% (p=0.002 n=6)
geomean                  69.37M        56.11M       -19.11%
```

dt committed:
sql/importer: avoid pointless rand read for imported uuids

Imported UUIDs are replaced with deterministic UUIDs to enable resuming imports.
Previously however, the in-memory UUID with this deterministic UUID was made
by making a random UUID -- reading random numbers from system entropy -- and
then replacing what it read with the deterministic numbers.

This skips the epensive and pointless init and just makes an empty struct for the
deterministic numbers to be set in.

Release note: none.
Epic: none.

dt committed:
sql/sem: add optional datum alloc to AdjustValueToType

In profiles, allocating new strings in this method is a significant CPU user during
imports. This adds an optional DatumAlloc that, if passed, is used and IMPORT passes
one.

Release note: none.
Epic: none.

dt committed:
sql/sem/tree: avoid pointless Round alloc in AdjustValueToType

Previously we unconditionally called ts.Round which would allocate a new
timestamp containing a rounded value. However, if the input timestamp is
already rounded this re-allocation is pointless and we can avoid it by
just returning the input timestamp if it is already rounded.

Release note: none.
Epic: none.

dt committed:
sql/row: SetTuple/ensureRawBytes via bufalloc

Profiles showed that during IMPORT, repeated calls to ensureRawBytes were a
significant source of allocations and CPU usage. This passes an allocator to
these paths.

Release note: none.
Epic: none.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/121395)
<!-- Reviewable:end -->

dt committed:
sql/importer: pool KVBatch slices

Release note: none.
Epic: none.

dt committed:
sql/row: avoid alloc per row for inserter func

Release note: none.
Epic: none.

dt committed:
sql/sem/tree: add Reset to DatumAlloc

This allows IMPORT to reuse datum allocations once it has encoded its row to bytes
since it knows the the lifetime of the datums allocated out of its pools. Only the
most common data types (string, int, decimal, uuid) are reused.

Release note: none.
Epic: none.

dt committed:
sql/row: add alloc pooling for []IndexEntry

Release note: none.
Epic: none.

dt committed:
sql/sem/tree: reduce AdjustType alloc for DDecimal

Release note: none.
Epic: none.

dt committed:
sql/importer: use b.N in benchmarkConvertToKVs

Release note: none.
Epic: none.

yuzefovich reviewed (commented):
Nice improvement!

> Reviewed 1 of 1 files at r1, 1 of 1 files at r2, 5 of 5 files at r3, 1 of 1 files at r4, 6 of 6 files at r5, 4 of 4 files at r6, 1 of 1 files at r7, 2 of 2 files at r8, 2 of 2 files at r9, 1 of 1 files at r10, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/121395#-:-NuQO84620kuQlyhplw0:bas2esp)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dt)

___
*[`-- commits` line 81 at r10](https://reviewable.io/reviews/cockroachdb/cockroach/121395#-NuQO2vK2x61RfoHWk9F):*
nit: seems like the last commit should be merged with the third one.
___
*[`pkg/roachpb/data.go` line 404 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/121395#-NuQIp4GC3OQPnfVuB78:-NuQI_f8CQyWbCzQC8pQ:b-y1a78t) ([raw file](https://github.com/cockroachdb/cockroach/blob/33917255b37749430de974f6def07061b4edea5a/pkg/roachpb/data.go#L404)):*
> ```Go
> func (v *Value) ensureRawBytesAlloc(size int, alloc *bufalloc.ByteAllocator) {
> 	if cap(v.RawBytes) < size {
> 		*alloc, v.RawBytes = alloc.Alloc(size, 0)
> ```

nit: add `/* extraCap */`.
___
*[`pkg/sql/row/row_converter.go` line 190 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/121395#-NuQLQEJD6aQMDcy7KvO:-NuQLQEJD6aQMDcy7KvP:b692ju5) ([raw file](https://github.com/cockroachdb/cockroach/blob/0eb26d74082ddd886dd95f8bba63d7734e201715/pkg/sql/row/row_converter.go#L190)):*
> ```Go
> 
> // KVBatch represents a batch of KVs generated from converted rows.
> type KVBatch struct {
> ```

nit: should the comment on this struct explicitly say that `Close` must be called once the "receiver" of this batch is done with it?

I think we're missing call to `Close` in `rowexec/bulk_row_writer.go:ingestLoop`.
___
*[`pkg/sql/row/row_converter.go` line 197 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/121395#-NuQJScRD2XqjeutnjEW:-NuQJScRD2XqjeutnjEX:bgqqolx) ([raw file](https://github.com/cockroachdb/cockroach/blob/0eb26d74082ddd886dd95f8bba63d7734e201715/pkg/sql/row/row_converter.go#L197)):*
> ```Go
> 	// Progress represents the fraction of the input that generated this row.
> 	Progress float32
> 	// KVs is the actual converted KV data.
> ```

nit: the comment is now off.
___
*[`pkg/sql/row/row_converter.go` line 209 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/121395#-NuQJkuF2IohiMJDj3NZ:-NuQJkuF2IohiMJDj3N_:bddgek6) ([raw file](https://github.com/cockroachdb/cockroach/blob/0eb26d74082ddd886dd95f8bba63d7734e201715/pkg/sql/row/row_converter.go#L209)):*
> ```Go
> func (k KVBatch) Close() {
> 	if k.pool != nil {
> 		k.KVs = k.KVs[:0]
> ```

Should we deeply reset each KV here (up to the `len`)?
___
*[`pkg/sql/row/row_converter.go` line 223 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/121395#-NuQKWP15PnAWAe2jdI2:-NuQKWP15PnAWAe2jdI3:b9b835o) ([raw file](https://github.com/cockroachdb/cockroach/blob/0eb26d74082ddd886dd95f8bba63d7734e201715/pkg/sql/row/row_converter.go#L223)):*
> ```Go
> 	KvCh     chan<- KVBatch
> 	KvBatch  KVBatch
> 	BatchCap int
> ```

nit: I think `BatchCap` is no longer used.
___
*[`pkg/sql/row/row_converter.go` line 548 at r8](https://reviewable.io/reviews/cockroachdb/cockroach/121395#-NuQNAPRDJ1lIQktre7i:-NuQNAPRDJ1lIQktre7j:bi0ssvf) ([raw file](https://github.com/cockroachdb/cockroach/blob/168e8d93817f094f95d7c371dbffc3aec1f7368a/pkg/sql/row/row_converter.go#L548)):*
> ```Go
> 	// Our datums don't live past the encoding into a KV so we can reset the alloc
> 	// pool with the same backing slice to reuse them rather than allocating new
> 	// datums for every call.
> ```

nit: this comment seems misaligned with `resetAllocBufs` call below - did you mean to defer that one here?
___
*[`pkg/sql/rowenc/index_encoding.go` line 1307 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/121395#-NuQN_J2AknGxu3elzD3:-NuQN_J2AknGxu3elzD4:b-as3vop) ([raw file](https://github.com/cockroachdb/cockroach/blob/1f6d02948fbc62c163ee2794aa0c9e8eea8dcde0/pkg/sql/rowenc/index_encoding.go#L1307)):*
> ```Go
> 	p := len(a.a)
> 	r := a.a[p : p : p+n]
> 	a.a = a.a[:p+n]
> ```

Shouldn't this be `a.a = a.a[p+n:]`?
___
*[`pkg/sql/sem/eval/cast.go` line 108 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121395#-NuQHBzn1nd99_vf72pi:-NuQHBzn1nd99_vf72pj:bh22l0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/792c9f6626bcedff802c09ca175c8b8b7508fae5/pkg/sql/sem/eval/cast.go#L108)):*
> ```Go
> 		return nil, err
> 	}
> 	return tree.AdjustValueToType(t, d, nil)
> ```

nit: add `/* alloc */`.
___
*[`pkg/sql/sem/tree/datum.go` line 6326 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/121395#-NuQHN0nDYzZWQQaQwTs:-NuQHN0nDYzZWQQaQwTt:b-3kqu4l) ([raw file](https://github.com/cockroachdb/cockroach/blob/792c9f6626bcedff802c09ca175c8b8b7508fae5/pkg/sql/sem/tree/datum.go#L6326)):*
> ```Go
> // the width of the value is wider than a single character. For this exception,
> // AdjustValueToType performs the truncation itself.
> func AdjustValueToType(typ *types.T, inVal Datum, alloc *DatumAlloc) (outVal Datum, err error) {
> ```

nit: mention that alloc is optional in the comment.
___
*[`pkg/sql/sem/tree/datum.go` line 6462 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/121395#-NuQHzWP2Ec_4AioL_vk:-NuQHzWP2Ec_4AioL_vl:bkn0of6) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f5904a87d9bedf8ed69d0e8017a343be4fd0afc/pkg/sql/sem/tree/datum.go#L6462)):*
> ```Go
> 	case types.TimeFamily:
> 		if in, ok := inVal.(*DTime); ok {
> 			return in.Round(TimeFamilyPrecisionToRoundDuration(typ.Precision())), nil
> ```

nit: let's apply the same optimization to Time, TimeTZ, and TimestampTZ types.
___
*[`pkg/sql/sem/tree/datum.go` line 6467 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/121395#-NuQHf_L1LJmp6FwVg5y:-NuQHf_L1LJmp6FwVg5z:babczg4) ([raw file](https://github.com/cockroachdb/cockroach/blob/6f5904a87d9bedf8ed69d0e8017a343be4fd0afc/pkg/sql/sem/tree/datum.go#L6467)):*
> ```Go
> 		if in, ok := inVal.(*DTimestamp); ok {
> 			p := TimeFamilyPrecisionToRoundDuration(typ.Precision())
> 			// If we don't need to round, skip calling it an allocating a new Datum.
> ```

nit: s/an allocating/and allocating/
___
*[`pkg/sql/importer/bench_test.go` line 92 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/121395#-NuQGMn4CHuDbFqOf8Sn:-NuQGMn4CHuDbFqOf8So:b-jrw8t4) ([raw file](https://github.com/cockroachdb/cockroach/blob/195bddcb297ac9ac2d347a7e36a1ddf64cd89ce1/pkg/sql/importer/bench_test.go#L92)):*
> ```Go
> 	var bytes int
> 	b.ResetTimer()
> 	for i := 0; i < b.N; i++ {
> ```

nit: it'd be good to backport the first commit (or just `b.N` change) to older branches to preserve meaningful comparison.


<!-- Sent from Reviewable.io -->

dt commented:
I don't think I'm going to get around to landing this as a combined PR. I'll pull out the easy bits and land them in their own changes.
